src/linker/Linker.Steps/ResolveFromAssemblyStep.cs

```diff
@@ -164,24 +176,31 @@ static void MarkType (LinkContext context, TypeDefinition type, RootVisibility r
 			if (type.HasNestedTypes)
 				foreach (var nested in type.NestedTypes)
 					MarkType (context, nested, rootVisibility);
+			if (type.HasInterfaces)
```

Comment:
I get this is the only change from this whole PR but I am not sure this is right. Why do you need to mark the base types explicitly here and not to do it in standard MarkStep?

Reply:
MarkStep won't mark the interfaces unless it sees that the type implementing it has been instantiated. We need some way to cause interface impls to be kept if they are considered part of the public surface area. Since we already have some logic in ResolveFromAssemblyStep to deal with marking public surface area based on visibility, I thought it made the most sense to do it here.

Reply:
I think we need to think about a different approach. You are trying to disable optimizations we are added and will keep adding to make linking smarter/better. With this approach, it will be constant work of undoing improvements in marking step (e.g. if we introduce base types reduction, or replacing interfaces with objects, etc).

Secondly, even this change is incomplete. What about cases where an interface is used as a constraint, or as attribute argument or typeof expression?

I think it'd be better to have for this mode a new step which replaces `ResolveFromAssemblyStep` and `MarkStep` as used now for `-r` called for example `MarkVisibleSurfaceStep`

Reply:
I'm not sure I understand the first concern - this doesn't disable any of the optimizations and they will still kick in for interface impls that aren't considered public surface area (or for all interface impls when linking apps using `-a`). For the scenario where we're linking individual corefx libraries, markstep needs to be told what to keep as public surface area.

> an interface is used as a constraint, or as attribute argument

Good point, I can work on adding those as long as we agree on the approach.

> typeof expression

I'm not sure that would be considered public surface area of a library, thoughts?

> a new step which replaces ResolveFromAssemblyStep and MarkStep as used now for -r

To me this seems like it would cause more duplication than the current approach - for example the logic for [rooting](https://github.com/mono/linker/blob/master/src/linker/Linker.Steps/ResolveFromAssemblyStep.cs#L97) exported types up-front is shared between `-r` and `-a`.

Reply:
My point is that when we add more optimizations which change metadata we'll have to undo them here and therefore would be easier to just have a different pipeline from this "mode".

>I'm not sure that would be considered public surface area of a library, thoughts?

That's exactly the point. I think nobody knows that right now and it's very hard to get that from the code. You are running linker as IL optimizer which I think needs more thinking about how to approach going forward.

> To me this seems like it would cause more duplication 

I don't think there will be any duplication, we can refactor any code as we need.

> for example the logic for rooting  exported types up-front is shared
We should never mark exported types upfront when in linking mode unless they are needed.

I am happy to talk about this in more details.

Reply:
I started trying the alternative we discussed in https://github.com/mono/linker/pull/793.